fly_io/datadog_checks/fly_io/check.py

```diff
@@ -0,0 +1,231 @@
+# (C) Datadog, Inc. 2024-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+from six.moves.urllib.parse import quote_plus
+
+from datadog_checks.base import OpenMetricsBaseCheckV2
+
+from .config_models import ConfigMixin
+from .constants import (
+    APP_COUNT_METRIC,
+    MACHINE_COUNT_METRIC,
+    MACHINE_CPUS_METRIC,
+    MACHINE_GPUS_METRIC,
+    MACHINE_MEM_METRIC,
+    MACHINE_SWAP_SIZE_METRIC,
+    MACHINE_UP_STATE,
+    MACHINES_API_UP_METRIC,
+    VOLUME_BLOCK_SIZE_METRIC,
+    VOLUME_BLOCKS_AVAIL_METRIC,
+    VOLUME_BLOCKS_FREE_METRIC,
+    VOLUME_BLOCKS_METRIC,
+    VOLUME_CREATED_METRIC,
+    VOLUME_CREATED_STATE,
+    VOLUME_ENCRYPTED_METRIC,
+    VOLUME_SIZE_METRIC,
+)
+from .metrics import METRICS, RENAME_LABELS_MAP
+
+
+class FlyIoCheck(OpenMetricsBaseCheckV2, ConfigMixin):
+    DEFAULT_METRIC_LIMIT = 0
+    __NAMESPACE__ = 'fly_io'
+
+    def __init__(self, name, init_config, instances):
+        super().__init__(name, init_config, instances)
+        self.org_slug = self.instance.get('org_slug')
+        self.machines_api_endpoint = self.instance.get('machines_api_endpoint')
+        self.tags = [f"fly_org:{self.org_slug}"]
+        match_string = self.instance.get('match_string', '{__name__=~".+"}')
+        encoded_match_string = quote_plus(match_string)
+        default_endpoint = f"https://api.fly.io/prometheus/{self.org_slug}/federate?match[]={encoded_match_string}"
+        self.openmetrics_endpoint = self.instance.get('openmetrics_endpoint', default_endpoint)
+        self.instance['openmetrics_endpoint'] = self.openmetrics_endpoint
+
+    def get_default_config(self):
+
+        return {
+            'openmetrics_endpoint': self.openmetrics_endpoint,
+            'metrics': [METRICS],
+            'rename_labels': RENAME_LABELS_MAP,
+            'hostname_label': 'instance',
+        }
+
+    def _get_app_status(self, app_name):
+        self.log.debug("Getting app status for %s", app_name)
+        app_details_endpoint = f"{self.machines_api_endpoint}/v1/apps/{app_name}"
+        response = self.http.get(app_details_endpoint)
+        try:
+            response.raise_for_status()
+            app = response.json()
+            app_status = app.get("status")
+            return app_status
+        except Exception:
+            self.log.info("Failed to collect app status for app %s", app_name)
+            return None
+
+    def _submit_machine_guest_metrics(self, guest, tags, machine_id):
+        self.log.debug("Getting machine guest metrics for %s", machine_id)
+        num_cpus = guest.get("cpus")
+        cpu_kind = guest.get("cpu_kind")
+        num_gpus = guest.get("gpus")
+        gpu_kind = guest.get("gpu_kind")
+        memory = guest.get("memory_mb")
+        if num_cpus is not None:
+            additional_tags = [f"cpu_kind:{cpu_kind}"]
+            self.gauge(MACHINE_CPUS_METRIC, value=num_cpus, tags=additional_tags + tags, hostname=machine_id)
```

Comment:
Nit. Are we okay with this creating a host in the infra list with the machine_id? I guess we do it already with vmware so it's not something out of the ordinary.

Reply:
Yes! We'll update the backend and frontend accordingly to make sure these are represented well.